### DISCUSSION TITLE
在xmake里添加python生成C++头文件/源文件并使用生成文件的方式

### DISCUSSION THREAD

**Discussion by Frandy** [xmake-io/xmake#3589] [Q&A] [answered]
在项目里需要根据一些文本文件，通过python生成头文件和源文件，然后在项目里使用。
xmake直接使用生成的头文件和源文件是比较方便，但是如果想把文本文件和python处理文件作为输入文件，怎么在xmake里控制需要的头文件/源文件的生成和使用呢？
有查到自定义rule, 以及lex/yacc的实现，不过还不是很清楚怎么写。希望有更接近的示例或对自定义rule更详细的指导。
举个例子

```
target("test")
      set_kind(binary)
      add_files("main.cpp", "src/*.cpp")
```

依赖的文本文件 token.txt, python文件 gencode.py, 使用python gencode.py token.txt 生成token.h, token.cpp, 需要怎么实现呢？

**Comment by waruqi** [comment, marked as answer]
lex / yacc 里面就最基础的 demo 了，直接参考里面的 https://github.com/xmake-io/xmake/blob/master/xmake/rules/lex_yacc/lex/xmake.lua

**Reply by Frandy** [reply]
谢谢您的回复.

我看到了lex/yacc的例子，有几个疑问，方便解释一下吗？
1. 生成的文件放在了target:autogendir(), 对于头文件，怎么添加include查找路径呢？ 
2. 是否需要单个输入文件只生成一个源文件？对于多个输入文件，这里的支持是否一样？

**Reply by waruqi** [reply]
> 谢谢您的回复.
> 
> 我看到了lex/yacc的例子，有几个疑问，方便解释一下吗？
> 
> 1. 生成的文件放在了target:autogendir(), 对于头文件，怎么添加include查找路径呢？

https://github.com/xmake-io/xmake/blob/45495538d52c211c7cacd741f727234b2b920373/xmake/rules/lex_yacc/yacc/xmake.lua#L27

> 2. 是否需要单个输入文件只生成一个源文件？对于多个输入文件，这里的支持是否一样？

根据 set_extensions 匹配，原本就是多文件，挨个处理

一次处理所有，走 before_buildcmd_files

**Reply by Frandy** [reply]
嗯，谢谢，学会了。
写了一个简单的例子，基本work了.
